### PR TITLE
[FLINK-19886][hive] Integrate file compaction to Hive connector

### PR DESCRIPTION
**Pull request by JingsongLi**
## What is the purpose of the change

Integrate file compaction to Hive connector

## Brief change log

- Introduce `HivePartitionUtils`
- Introduce HiveCompactReader.java 
- Integrate file compaction to HiveTableSink and refactor it.

## Verifying this change

`HiveSinkCompactionITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

### CODE REVIEW

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveCompactReader.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+			return new HiveSourceSplit("id", path, 0, len, new String[0], null, createPartition(path));
+		}
+
+		private HiveTablePartition createPartition(Path path) {
+			Map<String, Object> partitionSpec = new LinkedHashMap<>();
+			Map<String, DataType> nameToTypes = new HashMap<>();
+			for (int i = 0; i < fieldNames.length; i++) {
+				nameToTypes.put(fieldNames[i], fieldTypes[i]);
+			}
+			for (Map.Entry<String, String> entry : extractPartitionSpecFromPath(path).entrySet()) {
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveCompactReader.java:138]: Is it possible that partition spec cannot be extracted from path? E.g. when we write to a static partition and the partition location is different from the default one?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveSinkCompactionITCase.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+	@Override
+	@Before
+	public void init() throws IOException {
+		hiveCatalog = HiveTestUtils.createHiveCatalog();
+		tEnv().registerCatalog(hiveCatalog.getName(), hiveCatalog);
+		tEnv().useCatalog(hiveCatalog.getName());
+
+		// avoid too large parallelism lead to scheduler dead lock in streaming mode
+		tEnv().getConfig().getConfiguration().set(
+				HiveOptions.TABLE_EXEC_HIVE_INFER_SOURCE_PARALLELISM, false);
```
Comment (lirui-apache) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveSinkCompactionITCase.java:62]: I think only HiveTableSource performs parallelism inference. But this test won't use HiveTableSource, right?

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveCompactReader.java (hunk 1 above)
Comment (JingsongLi) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveCompactReader.java:138]: We don't support this, now the path must be come from `table.getSd()`, and the rule must be partitioned path.

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveSinkCompactionITCase.java (hunk 2 above)
Comment (JingsongLi) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/connectors/hive/HiveSinkCompactionITCase.java:62]: We are using. There is `select * from sink_table`

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveCompactReader.java (hunk 1 above)
Comment (lirui-apache) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveCompactReader.java]: Let's then log a jira to track this. Although not popular, it's still a valid use case in hive.

**Review by lirui-apache — APPROVED** [review]
LGTM

File: flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveCompactReader.java (hunk 1 above)
Comment (JingsongLi) [reply flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/read/HiveCompactReader.java]: https://issues.apache.org/jira/browse/FLINK-20040

**Pull request merged**